Tak.py

```diff
@@ -1,21 +1,152 @@
 from collections import deque, namedtuple
+from enum import Enum
+from logging import raiseExceptions
 
 GameState = namedtuple('GameState', 'to_move, board, pieces, moves')
 
+class pieceType(Enum):
+        """"Enum enables easy access to standardized piece characteristics"""
+        BLACK, WHITE, WALL,TILE,CAPSTONE = range(5)
+
+class Piece:
+    """Tak game piece represents color, type, and XYZ position of piece"""
+    def __init__(self, color, type, position):
+        self.color = color
+        self.type = type
+        self.position = [position[0],position[1],position[2]] #row, col, dequeIndex
+    
+    def __repr__(self):
+        rep = "B" if self.color == pieceType.BLACK else "W"
+        if self.type == pieceType.WALL:
+            rep += "W"
+        elif self.type == pieceType.TILE:
+            rep += "T"
+        else:
+            rep += "C"
+        return rep
 
 class Tak:
     """Tak game object which controls all game aspects."""
-
-    def __init__(self, board_length, num_stones, num_capstones):
+    def __init__(self, board_length, num_stones, num_capstones, board = []):
         """Set the initial game state."""
-        board = [board_length][board_length]
-        for i in range(0, board_length):
-            for j in range(0, board_length):
-                board[i][j] = deque()
+        self.board = board
+        if len(self.board) == 0:
+            for i in range(board_length):
+                self.board.append([])
+                for j in range(board_length):
+                    self.board[i].append(deque())
         self.white_pieces = [num_stones, num_capstones]
         self.black_pieces = [num_stones, num_capstones]
         self.placements = [(x, y) for x in range(0, board_length) for y in range(0, board_length)]
         self.white_moves = []
         self.black_moves = []
+        self.board_length = board_length
         moves = self.placements + self.white_moves
-        self.initial = GameState(to_move='W', board=board, pieces=self.white_pieces, moves=moves)
+        self.initial = GameState(to_move='W', board=self.board, pieces=self.white_pieces, moves=moves)
+
+    def terminal_test(self):
+        """Terminal States:\n\t
+        Board Win: all pieces of a player are used or board is covered\n\t
+        Road Win (Dragon): a move results in an orthogonally connected path 
+        of player tiles/capstone pieces belonging to a single player. If a move 
+        causes both players to have roads, the moving player wins."""
+        allPiecesPlaced = self.white_pieces == [0,0] or self.black_pieces == [0,0]
+        
+        fewestPieces = 1
+        #Checks for roads and board coverage
+        for j,row in enumerate(self.board):
```

Comment:
We go through all of the rows here, but are we getting through the columns? I'm not finding where that happens.

Reply:
Each row is a list of deques, line 58 iterates over each deque (i.e. board 'space', a rowXcol pairing) in the current list "row"

Reply:
I see it now